boot/freeldr/freeldr/arch/amd64/multiboot.S

```diff
@@ -0,0 +1,217 @@
+/*
+ * PROJECT:     FreeLoader
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Support for the Multiboot v1 specification.
+ * COPYRIGHT:   Copyright 1998-2002 Brian Palmer <brianp@sginet.com>
+ *              Copyright 2024 Daniel Victor <ilauncherdeveloper@gmail.com>
+ */
+
+#include <asm.inc>
+#include <arch/pc/x86common.h>
+#include <multiboot.h>
+
+/* Multiboot v1 support
+ *
+ * Allows freeldr to be loaded as a "multiboot kernel" by
+ * other boot loaders like GRUB.
+ * This code is not referenced from anywhere. GRUB searches
+ * for the header signature and uses the header to load it.
+ */
+
+#define MB_INFO_SIZE                60   /* sizeof(multiboot_info_t) */
+#define MB_INFO_FLAGS_OFFSET        0
+#define MB_INFO_BOOT_DEVICE_OFFSET  12
+#define MB_INFO_COMMAND_LINE_OFFSET 16
+#define MB_INFO_MMAP_LEN_OFFSET     44
+#define MB_INFO_MMAP_ADDR_OFFSET    48
+#define MB_MMAP_SIZE                480  /* 20 * sizeof(memory_map_t) - up to 20 entries */
+#define CMDLINE_SIZE                256
+
+/*
+ * We want to execute at FREELDR_BASE (to be compatible with
+ * bootsector loading), but GRUB only allows loading of
+ * multiboot kernels above 1MB. So we let GRUB load us
+ * there and then relocate ourself to FREELDR_BASE.
+ */
+#define INITIAL_BASE HEX(200000)
+
+
+#ifdef _USE_ML
+EXTERN __bss_start__:DWORD
+EXTERN __bss_end__:DWORD
+#endif
+
+
+#ifdef _USE_ML
+.MBDATA SEGMENT PUBLIC 'DATA'
+#endif
+
+    /* Align to 32 bits boundary */
+    .align 4
+
+    /* Multiboot header */
+MultibootHeader:
+    /* magic */
+    .long MULTIBOOT_HEADER_MAGIC
+    /* flags */
+    .long MULTIBOOT_HEADER_FLAGS
+    /* checksum */
+    .long -(MULTIBOOT_HEADER_MAGIC + MULTIBOOT_HEADER_FLAGS)
+    /* header_addr */
+    .long MultibootHeader + INITIAL_BASE - FREELDR_BASE
+    /* load_addr */
+    .long INITIAL_BASE
+    /* load_end_addr */
+    .long 0
+    /* bss_end_addr */
+    .long 0
+    /* entry_addr */
+    .long MultibootEntry + INITIAL_BASE - FREELDR_BASE
+
+#ifdef _USE_ML
+.MBDATA ENDS
+#endif
+
+
+.code32
+ASSUME ES:NOTHING, FS:NOTHING, GS:NOTHING
+
+MultibootEntry:
+    cld
+
+    /* Check for valid multiboot signature */
+    cmp eax, MULTIBOOT_BOOTLOADER_MAGIC
+    jne mbfail
+
+    /* Save multiboot info structure */
+    mov esi, ebx
+    mov edi, offset MultibootInfo + INITIAL_BASE - FREELDR_BASE
+    mov ecx, (MB_INFO_SIZE / 4)
+    rep movsd
+    mov dword ptr ds:[MultibootInfo + INITIAL_BASE - FREELDR_BASE + MB_INFO_MMAP_ADDR_OFFSET], 0
+    mov dword ptr ds:[_MultibootInfoPtr + INITIAL_BASE - FREELDR_BASE], offset MultibootInfo
+
```

Comment:
Does this code differ a lot from the one in the i386/multiboot.S file?
If not, what are its differences?

Reply:
No.

Reply:
Well, it seems it differs, since you've introduced (back) the amd64/multiboot.S file :joy:

You may want to investigate the usage of `#ifdef _M_IX86` (or its negative with `#ifndef`) for the few (??) changes.